src/Symfony/Component/Form/Form.php

```diff
@@ -550,10 +558,13 @@ public function submit($submittedData, bool $clearMissing = true)
             }
 
             // Hook to change content of the data submitted by the browser
-            if ($dispatcher->hasListeners(FormEvents::PRE_SUBMIT)) {
-                $event = new PreSubmitEvent($this, $submittedData);
-                $dispatcher->dispatch($event, FormEvents::PRE_SUBMIT);
-                $submittedData = $event->getData();
+            foreach ($thisModelDataAwareForms as $form) {
+                $dispatcher = $form->getConfig()->getEventDispatcher();
+                if ($dispatcher->hasListeners(FormEvents::PRE_SUBMIT)) {
+                    $event = new PreSubmitEvent($form, $submittedData);
```

Comment:
this change looks wrong to me. the submitted data will be in the nested structure corresponding to the form hierarchy. And in your patch, listeners on a child currently receiving the submitted data of the parent, thus needing to be aware of where they are nested.

Reply:
This feature is restricted to `inherit_data` option set to `true`. Passing the submitted data of the parent is exactly what we want to do to follow option's behavior. Or am I misunderstanding your point?

Reply:
the *submitted* data is not the same when you inherit data (because the submitted data has the nested structure where your child form exist in the names).

Reply:
Ok! Last commit should solve this.